cmd/coder/urls.go

```diff
@@ -33,55 +55,110 @@ var urlAccessLevel = map[string]string{
 	"PUBLIC":  "Anyone on the internet can access this link",
 }
 
-func portIsValid(port string) bool {
+func validatePort(port string) (int, error) {
 	p, err := strconv.ParseUint(port, 10, 16)
+	if err != nil {
+		flog.Error("Invalid port")
+		return 0, err
+	}
 	if p < 1 {
 		// port 0 means 'any free port', which we don't support
 		err = strconv.ErrRange
+		flog.Error("Port must be > 0")
+		return 0, err
 	}
-	if err != nil {
-		fmt.Println("Invalid port")
-	}
-	return err == nil
+	return int(p), nil
 }
 
 func accessLevelIsValid(level string) bool {
 	_, ok := urlAccessLevel[level]
 	if !ok {
-		fmt.Println("Invalid access level")
+		flog.Error("Invalid access level")
 	}
 	return ok
 }
 
+type listSubCmd struct {
+	outputFmt string
+}
+
+// Run gets the list of active devURLs from the cemanager for the
+// specified environment and outputs info to stdout.
+func (sub listSubCmd) Run(fl *pflag.FlagSet) {
+	envName := fl.Arg(0)
+	devURLs := urlList(envName)
+
+	switch sub.outputFmt {
+	case "human":
+		w := tabwriter.NewWriter(os.Stdout, 0, 0, 1, ' ', tabwriter.TabIndent)
+		for _, devURL := range devURLs {
+			fmt.Fprintf(w, "%s\t%d\t%s\n", devURL.URL, devURL.Port, devURL.Access)
+		}
+		err := w.Flush()
+		if err != nil {
+			flog.Fatal("failed to flush writer: %v", err)
```

Comment:
Checkout `internal/x/xtabwriter`. Should help make this section much cleaner. `secrets.go` has a good example.

Reply:
I've been thinking of adding support for a struct tag `human:"omit"` such that you could only print the fields that you want...

Reply:
Could be cool. What if users wanted more verbose vs. brief output -- would something like `human:"verbose"` and `human:"always"` or something allow multiple verbosity levels?

Reply:
As discussed via Slack I'll leave this for now but once xtabwriter supports marking fields to suppress output I will convert it to use that. (Just didn't want to output the `name` field since it's redundant, being part of the devurl itself).